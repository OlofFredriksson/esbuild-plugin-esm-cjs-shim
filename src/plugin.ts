import { type PluginBuild, type Plugin } from "esbuild";

export const shimPlugin = (): Plugin => ({
    name: "shimPlugin",
    setup(build: PluginBuild) {
        const options = build.initialOptions;

        if (!options.format) {
            throw new Error(
                `options.format needs to be defined in order to use plugin`
            );
        }
        if (options.format === "esm") {
            options.inject = [`${__dirname}/../shims/esm.mjs`];
            options.define = {
                ...options.define,
                ...{
                    require: "shimRequire",
                },
            };
        }

        if (options.format === "cjs") {
            options.inject = [`${__dirname}/../shims/cjs.js`];
            options.define = {
                ...options.define,
                ...{
                    "import.meta.url": "importMetaUrl",
                },
            };
        }
    },
});
