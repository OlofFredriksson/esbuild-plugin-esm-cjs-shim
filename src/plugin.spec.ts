import { build, type BuildOptions } from "esbuild";
import { shimPlugin } from "./plugin";

const defaultConfig: BuildOptions = {
    entryPoints: ["./test/fixtures/cjs.js"],
    bundle: true,
    write: false,
    plugins: [shimPlugin()],
    platform: "node",
    logLevel: "silent",
};

describe("shim", () => {
    let config: BuildOptions;
    beforeEach(() => {
        config = defaultConfig;
    });

    it("should throw error if format is missing", async () => {
        config = {
            ...config,
            entryPoints: ["./test/fixtures/cjs.js"],
        };

        await expect(build(config)).rejects.toThrow(Error);
    });

    it("should be able to transform commonJs to ESM", async () => {
        config = {
            ...config,
            entryPoints: ["./test/fixtures/cjs.js"],
            format: "esm",
        };
        const result = await build(config);
        expect(result.outputFiles?.[0].text).toMatchSnapshot();
    });

    it("should be able to transform esm to commonJs", async () => {
        config = {
            ...config,
            entryPoints: ["./test/fixtures/esm.mjs"],
            format: "cjs",
        };
        const result = await build(config);
        expect(result.outputFiles?.[0].text).toMatchSnapshot();
    });
});
