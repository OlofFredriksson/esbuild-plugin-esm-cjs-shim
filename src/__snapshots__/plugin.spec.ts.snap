// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shim should be able to transform commonJs to ESM 1`] = `
"// shims/esm.mjs
import { fileURLToPath } from "url";
import path from "path";
import { createRequire } from "node:module";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();
var __filename = /* @__PURE__ */ getFilename();
var shimRequire = /* @__PURE__ */ createRequire(import.meta.url);

// test/fixtures/cjs.js
console.log(__filename);
console.log(__dirname);
shimRequire.resolve("./cjs.js");
"
`;

exports[`shim should be able to transform esm to commonJs 1`] = `
""use strict";

// shims/cjs.js
var getImportMetaUrl = () => require("url").pathToFileURL(__filename).toString();
var importMetaUrl = /* @__PURE__ */ getImportMetaUrl();

// test/fixtures/esm.mjs
var import_path = require("path");
var import_url = require("url");
var __dirname = (0, import_path.dirname)((0, import_url.fileURLToPath)(importMetaUrl));
console.log(__dirname);
"
`;
